### PR TITLE
Support RTDB Emulator via FIREBASE_DATABASE_EMULATOR_HOST.

### PR DESCRIPTION
**Pull request by yuchenshi**
### Discussion

(Discussed internally.)

### Testing

I've manually ran the integration tests with and without the emulator and verified that they pass.
We may consider running the emulator-based integration test on Travis in a follow up PR.

### API Changes

No public API change except FakeCredential and the FIREBASE_DATABASE_EMULATOR_HOST env var.

### CODE REVIEW

File: firebase_admin/db.py [hunk 1]
```diff
@@ -768,46 +771,99 @@ class _DatabaseService(object):
...
+            if '//' in emulator_host:
+                raise ValueError(
+                    'Invalid {0}: "{1}". It must follow format "host:port".'.format(
+                        _EMULATOR_HOST_ENV_VAR, emulator_host))
+            use_fake_creds = True
+            host_override = emulator_host
+        else:
+            use_fake_creds = False
+            host_override = None
+        base_url, params = _DatabaseService._parse_db_url(db_url, host_override)
```
Comment (hiranya911) [inline firebase_admin/db.py]: If the databaseURL option has been set to an emulator URL, we will still try to use real credentials. Is that ok?

File: integration/conftest.py [hunk 2]
```diff
@@ -35,6 +37,9 @@ def _get_cert_path(request):
                      '"--cert" command-line option.')
 
 def integration_conf(request):
+    project_id = request.config.getoption('--project')
+    if project_id:
+        return credentials.FakeCredential(), project_id
```
Comment (hiranya911) [inline integration/conftest.py]: Would it be possible to move this logic to `integration/test_db.py`? There we can have something like:

```
# integration/test_db.py

def db_integration_conf():
    host_override = os.environ.get('FIREBASE_DATABASE_EMULATOR_HOST')
    if host_override:
        return db._FakeCredential(), 'fake-project-id'
    return conftest.integration_conf()
```

Then make sure integration tests use an App instance created using this test configuration.

This way we don't need a separate global `--project` flag, and the changes are contained to the test_db file.

File: firebase_admin/credentials.py [hunk 3]
```diff
@@ -199,3 +199,19 @@ def get_credential(self):
         Returns:
           google.auth.credentials.Credentials: A Google Auth credential instance."""
         return self._g_credential
+
+
+class FakeCredential(Base):
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply firebase_admin/credentials.py]: Done.

File: firebase_admin/db.py [hunk 4]
```diff
@@ -768,46 +771,99 @@ class _DatabaseService(object):
...
+
+        if self._auth_override:
+            params['auth_variable_override'] = self._auth_override
+
+        client_cache_key = (base_url, json.dumps(params, sort_keys=True), use_fake_creds)
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply firebase_admin/db.py]: Removed.

File: firebase_admin/db.py [hunk 5]
```diff
@@ -768,46 +771,99 @@ class _DatabaseService(object):
...
+        if client_cache_key not in self._clients:
+            credential = FakeCredential() if use_fake_creds else self._credential
+            client = _Client(credential.get_credential(), base_url, self._timeout, params)
+            self._clients[client_cache_key] = client
+        return self._clients[client_cache_key]
 
     @classmethod
-    def _validate_url(cls, url):
-        """Parses and validates a given database URL."""
+    def _parse_db_url(cls, url, host_override=None):
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply firebase_admin/db.py]: Done.

File: firebase_admin/db.py [hunk 6]
```diff
@@ -768,46 +771,99 @@ class _DatabaseService(object):
...
-            client = _Client(self._credential, base_url, self._auth_override, self._timeout)
-            self._clients[base_url] = client
-        return self._clients[base_url]
+    def get_client(self, db_url=None):
+        """Creates a client based on the db_url. Clients may be cached."""
+        if db_url is None:
+            db_url = self._db_url
+
+        emulator_host = os.environ.get(_EMULATOR_HOST_ENV_VAR)
+        if emulator_host:
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply firebase_admin/db.py]: Done.

File: firebase_admin/db.py (hunk 1 above)
Comment (yuchenshi) [reply firebase_admin/db.py]: Fixed. Right now all emulator URLs will use fake credentials.

File: integration/conftest.py (hunk 2 above)
Comment (yuchenshi) [reply integration/conftest.py]: Done.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -195,6 +195,18 @@ Now you can invoke the integration test suite as follows:
...
+### Emulator-based Integration Testing
+
+Some integration tests can run against emulators. This allows local testing
+without using real projects or credentials. For now, only the RTDB Emulator
+is supported.
+
+First, install the Firebase CLI, then run:
+
+```
+firebase emulators:exec --only database --project fake-project-id 'pytest integration/test_db.py'
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply CONTRIBUTING.md]: This `--project` flag is passed into `emulators:exec`. Unfortunately, it still needs a project to work.

File: firebase_admin/__init__.py [hunk 8]
```diff
@@ -263,6 +238,26 @@ def options(self):
...
+        variables.
+
+        Returns:
+            str: A project ID string or None.
+        """
+        if self._project_id is None:
+            try:
+                project_id = self._credential.project_id
+            except AttributeError:
+                project_id = None
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply firebase_admin/__init__.py]: The code below requires `project_id` to be bound.

File: firebase_admin/db.py [hunk 9]
```diff
@@ -768,46 +771,110 @@ class _DatabaseService(object):
...
+                        _EMULATOR_HOST_ENV_VAR, emulator_host))
+            self._emulator_host = emulator_host
+        else:
+            self._emulator_host = None
+
+    def get_client(self, db_url=None):
+        """Creates a client based on the db_url. Clients may be cached."""
+        if db_url is None:
+            db_url = self._db_url
+        base_url, params, use_fake_creds = \
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply firebase_admin/db.py]: I'd prefer to keep things conceptually separated. Also, in case we do want to support some params for production URLs (e.g. writeSizeLimit), we won't need to dramatically change the logic.

File: firebase_admin/db.py [hunk 10]
```diff
@@ -768,46 +771,110 @@ class _DatabaseService(object):
...
                 'URL string.'.format(url))
-        parsed = urllib.parse.urlparse(url)
-        if parsed.scheme != 'https':
-            raise ValueError(
-                'Invalid database URL: "{0}". Database URL must be an HTTPS URL.'.format(url))
-        elif not parsed.netloc.endswith('.firebaseio.com'):
+        parsed_url = urllib.parse.urlparse(url)
+        use_fake_creds = False
+        if parsed_url.netloc.endswith('.firebaseio.com'):
+            if parsed_url.scheme != 'https':
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply firebase_admin/db.py]: In reply to this and the comment below to make `_parse_emulator_url` throw -- the two errors need `url` instead of `parsed_url`. Is it worth it to pass both `(parsed_url, url)` in just for the sake of throwing errors?

File: integration/test_db.py [hunk 11]
```diff
@@ -25,6 +26,23 @@
...
+        return conftest.integration_conf(request)
+
+
+@pytest.fixture(autouse=True, scope='module')
+def default_app(request):
+    cred, project_id = integration_conf(request)
+    ops = {
+        'databaseURL' : 'https://{0}.firebaseio.com'.format(project_id),
+    }
+    return firebase_admin.initialize_app(cred, ops)
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply integration/test_db.py]: This actually is fine because the scope is `module` (i.e. `integration.test_db`) -- it will not touch other integration tests.

[the remark this answers is not in the document]
Comment (yuchenshi) [reply integration/test_db.py]: Oh, actually no, because this may initialize the global app again -- I see the issue now. Let me fix this.

[the remark this answers is not in the document]
Comment (yuchenshi) [reply integration/test_db.py]: Fixed.

File: firebase_admin/db.py [hunk 12]
```diff
@@ -768,46 +771,110 @@ class _DatabaseService(object):
...
             raise ValueError(
                 'Invalid database URL: "{0}". Database URL must be a valid URL to a '
                 'Firebase Realtime Database instance.'.format(url))
-        return 'https://{0}'.format(parsed.netloc)
+        if base_url == 'https://{0}.firebaseio.com'.format(namespace):
+            # namespace can be inferred from the base_url. No need for query params.
+            return base_url, {}, use_fake_creds
+        else:
+            # ?ns=foo is needed.
+            return base_url, {'ns': namespace}, use_fake_creds
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply firebase_admin/db.py]: I tried that but I found the return value `ns` really confusing, because it's not always the namespace name. It may be None if not needed, etc. and cannot be used for anything other than param. If we were to go with simplifying `params`, then I would actually move the param logic entirely out of the parsing method, and have something like:

```python
base_url, namespace = _parse_db_url()  # namespace will never be None

if base_url == 'https://{0}.firebaseio.com'.format(namespace):
    use_fake_creds = False
    params = {}
else:
    use_fake_creds = True
    params = {'ns': ns}
```

And we can further simply the parsing method. WDYT?

File: firebase_admin/__init__.py [hunk 13]
```diff
@@ -215,39 +215,14 @@ def __init__(self, name, credential, options):
...
-        if not project_id:
-            project_id = os.environ.get('GOOGLE_CLOUD_PROJECT', os.environ.get('GCLOUD_PROJECT'))
+        # Validate project_id if specified. Otherwise, we will try to infer
+        # from credential and then environment variables later when accessed.
+        project_id = self._options.get('projectId')
         if project_id is not None and not isinstance(project_id, six.string_types):
             raise ValueError(
                 'Invalid project ID: "{0}". project ID must be a string.'.format(project_id))
-        return project_id
+        self._project_id = project_id
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply firebase_admin/__init__.py]: Done. I'm not a huge fan of the tuple though so I made a flag for loading it only once. I hope you don't mind :)

File: firebase_admin/db.py [hunk 14]
```diff
@@ -768,46 +771,110 @@ class _DatabaseService(object):
...
+    @classmethod
+    def _parse_production_url(cls, parsed_url):
+        base_url = 'https://{0}'.format(parsed_url.netloc)
+        return base_url, parsed_url.netloc.split('.')[0]
+
+    @classmethod
+    def _parse_emulator_url(cls, parsed_url):
+        # Handle emulator URL like http://localhost:8080/?ns=foo-bar
+        query_ns = urllib.parse.parse_qs(parsed_url.query).get('ns')
+        if parsed_url.scheme in ['http', 'https']:
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply firebase_admin/db.py]: Uh I have no idea why I did this. Fixed.

File: firebase_admin/db.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (hiranya911) [reply firebase_admin/db.py]: I kind of like that. You can also get rid of the use_fake_creds by doing this:

```
base_url, namespace = _parse_db_url()  # namespace will never be None

if base_url == 'https://{0}.firebaseio.com'.format(namespace):
    credentials = self._credentials.get_credentials()
    params = {}
else:
    credentials = _EmulatorAdminCredentials()
    params = {'ns': namespace}
```

File: firebase_admin/__init__.py [hunk 15]
```diff
@@ -215,39 +215,15 @@ def __init__(self, name, credential, options):
...
-            except AttributeError:
-                pass
-        if not project_id:
-            project_id = os.environ.get('GOOGLE_CLOUD_PROJECT', os.environ.get('GCLOUD_PROJECT'))
+    @classmethod
+    def _validate_project_id(cls, project_id):
         if project_id is not None and not isinstance(project_id, six.string_types):
             raise ValueError(
                 'Invalid project ID: "{0}". project ID must be a string.'.format(project_id))
-        return project_id
```
Comment (yuchenshi) [inline firebase_admin/__init__.py]: We did not raise an exception when project_id is the empty string `""`, e.g. In cases like `options={projectId: ""}`. Is this intentional? Or shall I add a check for it? Or shall I normalize it to `None`?

File: firebase_admin/db.py [hunk 16]
```diff
@@ -768,46 +771,110 @@ class _DatabaseService(object):
...
-        return 'https://{0}'.format(parsed.netloc)
+        if base_url == 'https://{0}.firebaseio.com'.format(namespace):
+            # namespace can be inferred from the base_url. No need for query params.
+            return base_url, {}, use_fake_creds
+        else:
+            # ?ns=foo is needed.
+            return base_url, {'ns': namespace}, use_fake_creds
+
+    @classmethod
+    def _parse_production_url(cls, parsed_url):
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply firebase_admin/db.py]: Thanks, but no, thanks. I just tried that and it made testing a nightmare. I don't want to initialize a full `_DatabaseService` class with all the options and monkey-patch `os.environ` just to test the parsing logic.

File: firebase_admin/db.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (yuchenshi) [reply firebase_admin/db.py]: Yay, done!

File: firebase_admin/__init__.py (hunk 15 above)
Comment (hiranya911) [reply firebase_admin/__init__.py]: Not really intentional. I suppose because of how our lookup logic is implemented, `project_id` is empty only if the corresponding env variables are set to the empty string. We can continue to ignore that case, or if you like go ahead and normalize to None.

File: firebase_admin/db.py [hunk 17]
```diff
@@ -768,46 +771,110 @@ class _DatabaseService(object):
...
+        """
         if not url or not isinstance(url, six.string_types):
             raise ValueError(
                 'Invalid database URL: "{0}". Database URL must be a non-empty '
                 'URL string.'.format(url))
-        parsed = urllib.parse.urlparse(url)
-        if parsed.scheme != 'https':
+        parsed_url = urllib.parse.urlparse(url)
+        if parsed_url.netloc.endswith('.firebaseio.com'):
+            base_url, namespace = cls._parse_production_url(parsed_url, emulator_host)
```
Comment (hiranya911) [inline firebase_admin/db.py]: return

File: firebase_admin/db.py [hunk 18]
```diff
@@ -768,46 +771,110 @@ class _DatabaseService(object):
...
+        else:
+            base_url, namespace = cls._parse_emulator_url(parsed_url)
```
Comment (hiranya911) [inline firebase_admin/db.py]: return

File: integration/test_db.py [hunk 19]
```diff
@@ -25,11 +26,35 @@
...
+@pytest.fixture(scope='module')
+def app(request):
+    cred, project_id = integration_conf(request)
+    ops = {
+        'databaseURL' : 'https://{0}.firebaseio.com'.format(project_id),
+    }
+    return firebase_admin.initialize_app(cred, ops, name='integration-db')
+
+
+@pytest.fixture(scope='module', autouse=True)
```
Comment (hiranya911) [inline integration/test_db.py]: Why is this necessary?

File: tox.ini [hunk 20]
```diff
@@ -7,7 +7,9 @@
 envlist = py2,py3,pypy,cover
 
 [testenv]
-commands = pytest
+passenv =
+    FIREBASE_DATABASE_EMULATOR_HOST
```
Comment (hiranya911) [inline tox.ini]: What does this do exactly? We use this tox config during our releases. Will it be affected in any way?

File: integration/test_db.py (hunk 19 above)
Comment (yuchenshi) [reply integration/test_db.py]: Otherwise, the `default_app` in `conftest.py` will run (because it's `autouse`) and fail even if we only run the db integration tests with the emulator.

File: tox.ini (hunk 20 above)
Comment (yuchenshi) [reply tox.ini]: tox by default filters out most environment variables when running subprocesses like `tox`. This line basically tells tox that if `FIREBASE_DATABASE_EMULATOR_HOST` is set (e.g. `FIREBASE_DATABASE_EMULATOR_HOST=localhost:9000 tox`) then tox should pass along this env to pytest. This won't affect releases since we do not set `FIREBASE_DATABASE_EMULATOR_HOST` during releases.

File: firebase_admin/__init__.py (hunk 15 above)
Comment (yuchenshi) [reply firebase_admin/__init__.py]: Okay, I'm continuing to ignore that case.
